### PR TITLE
REFACTOR ALL THE APIS

### PR DESCRIPTION
**Pull request by FiloSottile**

### CODE REVIEW

File: internal/group/ge.go [hunk 1]
```diff
@@ -231,44 +232,30 @@ func (v *ProjectiveGroupElement) Zero() *ProjectiveGroupElement {
...
-	// returns a copy rather than smashing the receiver. It doesn't matter
-	// because it is always called on ephemeral intermediate values, but should
-	// fix.
-	var p, q ProjectiveGroupElement
-	var t0, t1 radix51.FieldElement
+func (v *ProjectiveGroupElement) DoubleZ1(u *ProjectiveGroupElement) *ProjectiveGroupElement {
+	var B, C, D, E, F radix51.FieldElement
 
-	p = *v
+	if u.Z.Equal(radix51.Zero) != 1 {
```
Comment (gtank) [inline internal/group/ge.go]: Shouldn't this be comparing to `radix51.One`?

File: internal/group/ge.go [hunk 2]
```diff
@@ -24,76 +31,79 @@ import (
...
 // Converts (x,y) to (X:Y:T:Z) extended coordinates, or "P3" in ref10. As
 // described in "Twisted Edwards Curves Revisited", Hisil-Wong-Carter-Dawson
 // 2008, Section 3.1 (https://eprint.iacr.org/2008/522.pdf)
 // See also https://hyperelliptic.org/EFD/g1p/auto-twisted-extended-1.html#addition-add-2008-hwcd-3
 func (v *ExtendedGroupElement) FromAffine(x, y *big.Int) {
-	field.FeFromBig(&v.X, x)
-	field.FeFromBig(&v.Y, y)
-	field.FeMul(&v.T, &v.X, &v.Y)
-	field.FeOne(&v.Z)
+	v.X.FromBig(x)
```
[the remark this answers is not in the document]
Comment (gtank) [reply internal/group/ge.go]: This is now https://github.com/gtank/ed25519/issues/9

File: ed25519.go [hunk 3]
```diff
@@ -125,10 +122,10 @@ func (curve ed25519Curve) ScalarMult(x1, y1 *big.Int, k []byte) (x, y *big.Int)
 		var b = int32((s[i/8] >> uint(i&7)) & 1)
 		if b == 0 {
 			r1.Add(&r0, &r1)
-			r0.Double()
+			r0.Double(&r0)
 		} else {
 			r0.Add(&r0, &r1)
-			r1.Double()
+			r1.Double(&r0)
```
[the remark this answers is not in the document]
Comment (FiloSottile) [reply ed25519.go]: FYI GitHub now has suggested changes in the PR flow.

image

File: internal/group/ge.go [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
 
 import (
 	"math/big"
 
-	field "github.com/gtank/ed25519/internal/radix51"
+	"github.com/gtank/ed25519/internal/radix51"
 )
 
+// D is a constant in the curve equation.
+var D = &radix51.FieldElement{929955233495203, 466365720129213,
```
[the remark this answers is not in the document]
Comment (FiloSottile) [reply internal/group/ge.go]: "anyone who wants to use this code with different curve parameters" is a cursed statement, but I'll derive `twoD` from `D`. This way we can keep the `twoD` definition close to where it's used.

File: internal/group/ge.go [hunk 5]
```diff
@@ -126,40 +136,33 @@ func (v *ExtendedGroupElement) Double() *ExtendedGroupElement {
...
 
 	// C ← 2*Z1^2
-	field.FeSquare(&C, &v.Z)
-	field.FeAdd(&C, &C, &C) // TODO should probably implement FeSquare2
+	C.Square(&v.Z)
+	C.Add(&C, &C) // TODO should probably implement FeSquare2
 
 	// D ← -1*A
-	field.FeNeg(&D, &A) // implemented as substraction
+	D.Neg(&A) // implemented as substraction
```
[the remark this answers is not in the document]
Comment (FiloSottile) [reply internal/group/ge.go]: Done.

File: internal/group/ge.go (hunk 1 above)
Comment (FiloSottile) [reply internal/group/ge.go]: It should, but now the question is why didn't it panic?

And the answer is that `(*ProjectiveGroupElement).FromAffine` sets Z to 0, but nothing uses it before it's set by `DoubleZ1`. Fixed.

File: internal/radix51/fe.go [hunk 6]
```diff
@@ -350,19 +360,19 @@ func (v *FieldElement) Equal(u *FieldElement) int {
 
 // Select sets v to a if cond == 1, and to b if cond == 0.
 // v, a and b are allowed to overlap.
-func (v *FieldElement) Select(a, b *FieldElement, cond int) {
+func (v *FieldElement) Select(a, b *FieldElement, cond int) *FieldElement {
 	m := uint64(cond) * 0xffffffffffffffff
```
[the remark this answers is not in the document]
Comment (FiloSottile) [reply internal/radix51/fe.go]: Done.

**Comment by FiloSottile** [comment]
> These conversion functions actually look like the caller is supposed to do the allocation, but the commit message says 'callee'

Ah, yes, that's a typo. Fixed.

**Pull request merged**